src/hera/workflows/_runner/script_annotations_util.py

```diff
@@ -99,30 +107,37 @@ def get_annotated_artifact_value(artifact_annotation: Artifact) -> Union[Path, A
     the `kwargs` or the function parameter name.
     """
     if artifact_annotation.output:
-        if artifact_annotation.path:
-            path = Path(artifact_annotation.path)
-        else:
-            path = _get_outputs_path(artifact_annotation)
-        # Automatically create the parent directory (if required)
-        path.parent.mkdir(parents=True, exist_ok=True)
-        return path
+        return get_annotated_output_artifact(artifact_annotation)
 
     if not artifact_annotation.path:
         # Path is added to the spec automatically when built. As it isn't present in the annotation itself,
         # we need to add it back in for the runner.
         artifact_annotation.path = artifact_annotation._get_default_inputs_path()
 
-    if artifact_annotation.loader == ArtifactLoader.json.value:
-        path = Path(artifact_annotation.path)
-        with path.open() as f:
-            return json.load(f)
+    actual_type = unwrap_annotation(func_param_annotation)
 
-    if artifact_annotation.loader == ArtifactLoader.file.value:
-        path = Path(artifact_annotation.path)
+    if artifact_annotation.loader is None:
+        if os.environ.get("hera__artifact_path_as_string", None) is not None and issubclass(actual_type, str):
+            warnings.warn(
+                "The artifact path is being passed as a string. This is deprecated and will be removed in future versions. "
+                "Please use a Path type for the function parameter.",
+            )
+            return artifact_annotation.path
+
+        if issubclass(actual_type, Path):
+            return artifact_annotation.path
+
+    path = Path(artifact_annotation.path)
+    if artifact_annotation.loader == ArtifactLoader.file:
         return path.read_text()
 
-    if artifact_annotation.loader is None:
-        return artifact_annotation.path
+    if artifact_annotation.loader == ArtifactLoader.json or artifact_annotation.loader is None:
+        with path.open() as f:
+            return json.load(f)
+
+    if isinstance(artifact_annotation.loader, Callable):  # type: ignore
+        loaded_value = artifact_annotation.loader(path.read_text())
+        return loaded_value
```

Comment:
What about codecs that want to work on `bytes`?

Reply:
If we renamed the attributes `loads` and `dumps` then we could have `load` and `dump` that take Paths.

Reply:
Sounds good – we should use `loads` and `dumps` in Parameter as well then

Reply:
I chose to use `bytes` rather than `Path` to mirror the `str` behaviour. Also then we have the user having to pick the `Path` to write to for the `dump` function.